### PR TITLE
Optimized fast atlas merging

### PR DESCRIPTION
**Pull request by Stannislav**
The new merging is impemented in the `atlannot.merge` module.

Test using these scripts
* `prep-data.py` make small versions of original merged atlas volumes for faster testing
* `check-{coarse,fine}.py`: check the merging on the small atlases
* `check-{coarse,fine}-real.py`: check the merging on the full volume (slower)

<details>
<summary>prep-data.py</summary>

```python
import logging
import pathlib
import sys

import nrrd
import numpy as np

logger = logging.getLogger(__name__)


def prep_atlas(filename):
    data, header = nrrd.read(filename)
    return data[::3, ::3, ::3]


def main() -> int:
    logger.info("Preparing CCFv2")
    ccfv2_small = prep_atlas("data/annotation_25_2011_full.nrrd")
    logger.info("Preparing CCFv3")
    ccfv3_small = prep_atlas("data/annotation_25_2017.nrrd")

    out_dir = pathlib.Path(__file__).parent
    logger.info("Saving CCFv2-small")
    np.save(out_dir / "ccfv2-small.npy", ccfv2_small)
    logger.info("Saving CCFV3-small")
    np.save(out_dir / "ccfv3-small.npy", ccfv3_small)

    return 0


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    sys.exit(main())
```
</details>

<details>
<summary>check-coarse.py</summary>

```python
import json
import logging
import pathlib
import sys
import time

import numpy as np

logger = logging.getLogger(__name__)


def main() -> int:
    logger.info("Imports")
    from atlannot.atlas.region_meta import RegionMeta
    from atlannot.merge import coarse
    from atlannot.merge_original.coarse import coarse_merge

    script_dir = pathlib.Path(__file__).parent
    logger.info("Loading CCFv2-small")
    ccfv2_small = np.load(script_dir / "ccfv2-small.npy")
    logger.info("Loading CCFv3-small")
    ccfv3_small = np.load(script_dir / "ccfv3-small.npy")

    logger.info("Loading brain regions")
    with open("data/brain_regions.json") as fp:
        brain_regions_data = json.load(fp)
        brain_regions = brain_regions_data["msg"][0]
    region_meta = RegionMeta.from_root_region(brain_regions)

    logger.info("Running fast merging")
    start = time.perf_counter()
    merged_fast = coarse.merge(ccfv2_small, ccfv3_small, region_meta)
    time_new = time.perf_counter() - start

    logger.info("Running old merging")
    start = time.perf_counter()
    merged = coarse_merge(ccfv2_small, ccfv3_small, brain_regions)
    time_old = time.perf_counter() - start

    logger.info("Performance:")
    logger.info("Old: %f", time_old)
    logger.info("New: %f", time_new)

    logger.info("Checking results")
    if np.array_equal(merged, merged_fast):
        logger.info("Passed!")
        return 0
    else:
        logger.error("Fail!")
        return 1


if __name__ == "__main__":
    formatter = logging.Formatter(
        fmt="%(asctime)s %(name)s (%(levelname).1s) %(message)s",
        datefmt="%H:%M:%S",
    )
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)

    root_logger = logging.getLogger()
    root_logger.setLevel(logging.INFO)
    root_logger.addHandler(handler)

    sys.exit(main())
```
</details>

<details>
<summary>check-fine.py</summary>

```python
import json
import logging
import pathlib
import sys
import time

import numpy as np

logger = logging.getLogger(__name__)


def main() -> int:
    logger.info("Imports")
    from atlannot.atlas.region_meta import RegionMeta
    from atlannot.merge import fine
    from atlannot.merge_original.fine import fine_merge

    script_dir = pathlib.Path(__file__).parent
    logger.info("Loading CCFv2-small")
    ccfv2_small = np.load(script_dir / "ccfv2-small.npy")
    logger.info("Loading CCFv3-small")
    ccfv3_small = np.load(script_dir / "ccfv3-small.npy")

    logger.info("Loading brain regions")
    with open("data/brain_regions.json") as fp:
        brain_regions_data = json.load(fp)
        brain_regions = brain_regions_data["msg"][0]
    region_meta = RegionMeta.from_root_region(brain_regions)

    logger.info("Running fast merging")
    start = time.perf_counter()
    merged_fast = fine.merge(ccfv2_small, ccfv3_small, region_meta)
    time_new = time.perf_counter() - start

    logger.info("Running old merging")
    start = time.perf_counter()
    merged = fine_merge(ccfv2_small, ccfv3_small, brain_regions)
    time_old = time.perf_counter() - start

    logger.info("Performance:")
    logger.info("Old: %f", time_old)
    logger.info("New: %f", time_new)

    logger.info("Checking results")
    if np.array_equal(merged, merged_fast):
        logger.info("Passed!")
        return 0
    else:
        logger.error("Fail!")
        return 1


if __name__ == "__main__":
    formatter = logging.Formatter(
        fmt="%(asctime)s %(name)s (%(levelname).1s) %(message)s",
        datefmt="%H:%M:%S",
    )
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)

    root_logger = logging.getLogger()
    root_logger.setLevel(logging.INFO)
    root_logger.addHandler(handler)

    sys.exit(main())
```
</details>

<details>
<summary>check-coarse-real.py</summary>

```python
import json
import logging
import pathlib
import sys
import time

import nrrd
import numpy as np

logger = logging.getLogger(__name__)


def main() -> int:
    logger.info("Imports")
    from atlannot.atlas.region_meta import RegionMeta
    from atlannot.merge import coarse

    logger.info("Loading original atlases")
    ccfv2, _ = nrrd.read("data/annotation_25_2011_full.nrrd")
    ccfv3, _ = nrrd.read("data/annotation_25_2017.nrrd")

    logger.info("Loading merged atlases")
    ccfv2_merged_model, _ = nrrd.read("data/ccfv2_atlas_coarse.nrrd")
    ccfv3_merged_model, _ = nrrd.read("data/ccfv3_atlas_coarse.nrrd")

    logger.info("Loading brain regions")
    with open("data/brain_regions.json") as fp:
        brain_regions = json.load(fp)
    region_meta = RegionMeta.from_root_region(brain_regions["msg"][0])

    logger.info("Start merging")
    start = time.perf_counter()
    ccfv2_merged, ccfv3_merged = coarse.merge(ccfv2, ccfv3, region_meta)
    duration = time.perf_counter() - start
    logger.info("Performance: %.1fs", duration)

    logger.info("Saving results")
    out_dir = pathlib.Path(__file__).parent
    np.save(out_dir / "ccfv2-merged-coarse.npy", ccfv2_merged)
    np.save(out_dir / "ccfv3-merged-coarse.npy", ccfv3_merged)

    logger.info("Checking results")
    check_v2 = np.array_equal(ccfv2_merged, ccfv2_merged_model)
    check_v3 = np.array_equal(ccfv3_merged, ccfv3_merged_model)
    if check_v2 and check_v3:
        logger.info("Everything matches!")
        return 0
    else:
        if not check_v2:
            logger.error("CCFv2 doesn't match :(")
        if not check_v3:
            logger.error("CCFv3 doesn't match :(")
        return 1


if __name__ == "__main__":
    formatter = logging.Formatter(
        fmt="%(asctime)s %(name)s (%(levelname).1s) %(message)s",
        datefmt="%H:%M:%S",
    )
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)

    root_logger = logging.getLogger()
    root_logger.setLevel(logging.INFO)
    root_logger.addHandler(handler)

    sys.exit(main())
```
</details>

<details>
<summary>check-fine-real.py</summary>

```python
import json
import logging
import pathlib
import sys
import time

import nrrd
import numpy as np

logger = logging.getLogger(__name__)


def main() -> int:
    logger.info("Imports")
    from atlannot.atlas.region_meta import RegionMeta
    from atlannot.merge import fine

    logger.info("Loading original atlases")
    ccfv2, _ = nrrd.read("data/annotation_25_2011_full.nrrd")
    ccfv3, _ = nrrd.read("data/annotation_25_2017.nrrd")

    logger.info("Loading merged atlases")
    ccfv2_merged_model, _ = nrrd.read("data/ccfv2_atlas_fine.nrrd")
    ccfv3_merged_model, _ = nrrd.read("data/ccfv3_atlas_fine.nrrd")

    logger.info("Loading brain regions")
    with open("data/brain_regions.json") as fp:
        brain_regions = json.load(fp)
    region_meta = RegionMeta.from_root_region(brain_regions["msg"][0])

    logger.info("Start merging")
    start = time.perf_counter()
    ccfv2_merged, ccfv3_merged = fine.merge(ccfv2, ccfv3, region_meta)
    duration = time.perf_counter() - start
    logger.info("Performance: %.1fs", duration)

    logger.info("Saving results")
    out_dir = pathlib.Path(__file__).parent
    np.save(out_dir / "ccfv2-merged-fine.npy", ccfv2_merged)
    np.save(out_dir / "ccfv3-merged-fine.npy", ccfv3_merged)

    logger.info("Checking results")
    check_v2 = np.array_equal(ccfv2_merged, ccfv2_merged_model)
    check_v3 = np.array_equal(ccfv3_merged, ccfv3_merged_model)
    if check_v2 and check_v3:
        logger.info("Everything matches!")
        return 0
    else:
        if not check_v2:
            logger.error("CCFv2 doesn't match :(")
        if not check_v3:
            logger.error("CCFv3 doesn't match :(")
        return 1


if __name__ == "__main__":
    formatter = logging.Formatter(
        fmt="%(asctime)s %(name)s (%(levelname).1s) %(message)s",
        datefmt="%H:%M:%S",
    )
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)

    root_logger = logging.getLogger()
    root_logger.setLevel(logging.INFO)
    root_logger.addHandler(handler)

    sys.exit(main())
```
</details>

### CODE REVIEW

**Review by EmilieDel — APPROVED** [review]
I don't understand every technical detail but it seems fine. Thanks a lot for doing all of this.

File: src/atlannot/merge/common.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        An arbitrary numpy array.
+    old_value
+        The value to replace.
+    new_value
+        The new value that replaces the old one.
+    """
+    array[array == old_value] = new_value
+
+
+def atlas_remap(
```
Comment (EmilieDel) [inline src/atlannot/merge/common.py:37]: Just thinking about this now, maybe we could replace the remapping of labels with that function instead ? I know we use it in `ants.transform` and it is a bit slow. What do you think ? Should we create a ticket for this ?

File: src/atlannot/atlas/region_meta.py [hunk 2]
```diff
@@ -75,6 +78,132 @@ def ids_at_level(self, level):
...
+        ----------
+        region_id
+            The region ID in question.
+
+        Returns
+        -------
+        int
+            The region ID of the parent.
+        """
+        return self.parent_id.get(region_id)
```
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atlannot/atlas/region_meta.py:111]: Nice, new change: e393290

File: src/atlannot/atlas/region_meta.py [hunk 3]
```diff
@@ -75,6 +78,132 @@ def ids_at_level(self, level):
...
+            The region ID in question.
+
+        Yields
+        ------
+        int
+            The region ID of a child region.
+        """
+        for child_id, parent_id in self.parent_id.items():
+            if parent_id == region_id:
+                yield child_id
```
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atlannot/atlas/region_meta.py:132]: Good point! I extended the docstring: 68735e4

File: src/atlannot/merge/coarse.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    v2_to = v2_from.copy()
+    v3_from = np.unique(ccfv3)
+    v3_to = v3_from.copy()
+
+    logger.info("First loop")
+    unique_v2 = set(v2_to)
+    unique_v3 = set(v3_to)
+    ids_to_correct = unique_v2 - unique_v3
+    for id_ in ids_to_correct:
+        if rm.is_leaf(id_) and id_ not in unique_v3 and rm.parent(id_) in unique_v3:
```
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atlannot/merge/coarse.py]: You're totally right. It was like that in the original algorithm, and I didn't change the logic. But I removed these redundant checks now: d4ac694

File: src/atlannot/merge/fine.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+    logger.info("Collecting all CCFv2 and CCFv3 region IDs")
+    all_v2_region_ids: set = rm.collect_ancestors(v2_to)
+    all_v3_region_ids: set = rm.collect_ancestors(v3_to)
+
+    logger.info("First for-loop correction")
+    unique_v2 = set(v2_to)
+    unique_v3 = set(v3_to)
+    ids_to_correct = unique_v2 - unique_v3
+    for id_ in ids_to_correct:
+        if rm.is_leaf(id_) and id_ not in unique_v3 and rm.parent(id_) in unique_v3:
```
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atlannot/merge/fine.py]: See above

File: src/atlannot/merge/common.py (hunk 1 above)
Comment (Stannislav) [reply src/atlannot/merge/common.py:37]: Yes, very good point! Maybe we can do that in a separate issue?

File: data/atlas_merge.py [hunk 6]
```diff
@@ -18,7 +18,7 @@
 
 import nrrd
 
-from atlannot.atlas_merge import coarse_merge, fine_merge
+from atlannot.merge_original import coarse_merge, fine_merge
```
[the remark this answers is not in the document]
Comment (Stannislav) [reply data/atlas_merge.py:21]: Yes, I agree! Maybe we can create a new issue for that?

File: src/atlannot/atlas/region_meta.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atlannot/atlas/region_meta.py:111]: Thanks :)

File: src/atlannot/merge/coarse.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atlannot/merge/coarse.py]: Perfect!

File: src/atlannot/merge/common.py (hunk 1 above)
Comment (EmilieDel) [reply src/atlannot/merge/common.py:37]: Sounds a good plan :)

File: data/atlas_merge.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (EmilieDel) [reply data/atlas_merge.py:21]: Sounds good for me :)

File: src/atlannot/atlas/region_meta.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atlannot/atlas/region_meta.py:132]: 👍

File: src/atlannot/merge/common.py (hunk 1 above)
Comment (Stannislav) [reply src/atlannot/merge/common.py:37]: Created #23

File: data/atlas_merge.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Stannislav) [reply data/atlas_merge.py:21]: Created #24

**Pull request merged**